File: app/src/main/java/nl/ngti/androscope/test/MainActivity.java
package nl.ngti.androscope.test;

import android.Manifest;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.button_grant_storage_permission).setOnClickListener(this);

        addSamplePreferences();
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.button_grant_storage_permission) {
            requestStoragePermission();
        }
    }

    private void addSamplePreferences() {
        SharedPreferences prefs = getApplicationContext().getSharedPreferences("sample", Context.MODE_PRIVATE);
        if (!prefs.contains("sample_key")) {
            prefs.edit().putString("sample_key", "Sample Value").apply();
        }
    }

    private void requestStoragePermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 999);
        }
    }
}

File: app/src/main/java/nl/ngti/androscope/test/MainActivity.kt
package nl.ngti.androscope.test

import android.Manifest
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import nl.ngti.androscope.AndroscopeActivity
import nl.ngti.androscope.test.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        ActivityMainBinding.inflate(layoutInflater).apply {
            setContentView(root)
            bindViews()
        }

        addSamplePreferences()
    }

    private fun ActivityMainBinding.bindViews() {
        openAndroscopeActivity.setOnClickListener {
            openAndroscopeActivity()
        }
        buttonGrantStoragePermission.setOnClickListener {
            requestPermissions(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        }
        buttonGrantContactsPermission.setOnClickListener {
            requestPermissions(Manifest.permission.READ_CONTACTS)
        }
    }

    private fun addSamplePreferences() {
        getSharedPreferences("sample", Context.MODE_PRIVATE).run {
            if (!contains("sample_key")) {
                edit().putString("sample_key", "Sample Value").apply()
            }
        }
    }

    private fun requestPermissions(vararg permissions: String) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(permissions, 999)
        }
    }

    private fun openAndroscopeActivity() {
        Intent(this, AndroscopeActivity::class.java).also {
            startActivity(it)
        }
    }
}
